bin/main.ml

```diff
@@ -80,83 +25,199 @@ let gen_output ast f = function
-    match (k, v) with
-    | Syntax.PrintUsage, _ -> usage () |> print_endline
-    | Syntax.ShowPragmas, _ -> Syntax.show_pragmas pragmas |> print_endline
-    | Syntax.NestedProtocols, _ ->
-        if Option.is_some !fsm then
-          Err.uerr (Err.IncompatibleFlag ("fsm", Syntax.show_pragma k))
+let main file enumerate verbose go_path out_dir project fsm gencode_ocaml
+    gencode_monadic_ocaml gencode_go =
+  let process_pragmas (pragmas : Syntax.pragmas) : unit =
+    let process_global_pragma (k, v) =
+      match (k, v) with
+      | Syntax.PrintUsage, _ -> ()
+      | Syntax.ShowPragmas, _ -> Syntax.show_pragmas pragmas |> print_endline
+      | Syntax.NestedProtocols, _ ->
+          if Option.is_some fsm then
+            Err.uerr (Err.IncompatibleFlag ("fsm", Syntax.show_pragma k))
+    in
+    List.iter ~f:process_global_pragma pragmas
   in
-  List.iter ~f:process_global_pragma pragmas
-
-let generate_code (ast : Syntax.scr_module) gencode =
-  let gen_ocaml role protocol =
-    Lib.generate_ocaml_code ~monad:false ast ~protocol ~role
-  in
-  let gen_go protocol =
-    match !out_dir with
-    | Some out_dir ->
-        Lib.generate_go_code ast ~protocol ~out_dir ~go_path:!go_path
-    | None ->
-        Err.UserError
-          (Err.MissingFlag
-             ( "out-dir"
-             , "This flag must be set in order to generate go implementation"
-             ))
-        |> raise
-  in
-  match gencode with
-  | Some (role, protocol) ->
-      let impl =
-        if Pragma.nested_protocol_enabled ast then gen_go protocol
-        else gen_ocaml role protocol
-      in
-      print_endline impl
-  | None -> ()
-
-let run filename verbose enumerate proj fsm gencode =
   try
-    let ast = process_file filename Lib.parse in
+    let ast = process_file file Lib.parse in
     process_pragmas ast.pragmas ;
     Lib.validate_exn ast ~verbose ;
-    if enumerate then
-      Lib.enumerate ast
-      |> List.map ~f:(fun (n, r) ->
-             RoleName.user r ^ "@" ^ ProtocolName.user n)
-      |> String.concat ~sep:"\n" |> print_endline
-    else () ;
-    gen_output ast
-      (fun ast protocol role ->
-        Lib.project_role ast ~protocol ~role |> Ltype.show)
-      proj ;
-    gen_output ast
-      (fun ast protocol role ->
-        Lib.generate_fsm ast ~protocol ~role |> snd |> Efsm.show)
-      fsm ;
-    generate_code ast gencode ;
-    true
+    let () =
+      if enumerate then
+        Lib.enumerate ast
+        |> List.map ~f:(fun (n, r) ->
+               RoleName.user r ^ "@" ^ ProtocolName.user n)
+        |> String.concat ~sep:"\n" |> print_endline
+    in
+    let () =
+      gen_output ast
+        (fun ast protocol role ->
+          Lib.project_role ast ~protocol ~role |> Ltype.show)
+        project
+    in
+    let () =
+      gen_output ast
+        (fun ast protocol role ->
+          Lib.generate_fsm ast ~protocol ~role |> snd |> Efsm.show)
+        fsm
+    in
+    let () =
+      match gencode_ocaml with
```

Comment:
Very optional, but can we consider a form of the maybe monad here? this is a bit...

I would not mind using let* and rising the minimum OCaml version as needed.

Reply:
You mean replace it with Option.iter?

Reply:
I didn't know of Option.iter, but that is a possibility yes.

Reply:
I like it.